cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,509 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <initializer_list>
+#include <limits>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace internal {
+
+#if __cplusplus >= 201703L
+using std::launder;
+#else
+// TODO factor out from result.h
+
+// template <class T>
+// constexpr T* launder(T* p) noexcept {
+//   return p;
+// }
+#endif
+
+template <typename ValueType, typename PointerType, typename ReferenceType>
+class VectorIterator {
```

Comment:
I'm not sure this is worthwhile; `std::iterator_traits` is already specialized for `int*` and `const int*` so I think you could just `using iterator = int*; using const_iterator = const int*;` below

Reply:
Wow, this is very couple if it works! Will check on CI.